docs/contributors/code/coding-guidelines.md

```diff
@@ -136,10 +136,317 @@ export { __unstableDoTerribleAwfulAction } from './api';
+
+    // ...
+}
+```
+
+##### Experimental functions, classes, and variables
+
+```js
+// In packages/package1/index.js:
+import { lock } from './experiments';
+
+export const experiments = {};
+/* Attach private data to the exported object */
+lock(experiments, {
+	__experimentalCallback: function() {},
+	__experimentalReactComponent: function ExperimentalComponent() { return <div/>; },
+	__experimentalClass: class Experiment{},
+	__experimentalVariable: 5,
+});
+
+
+// In packages/package2/index.js:
+import { experiments } from '@wordpress/package1';
+import { unlock } from './experiments';
+
+const {
+	__experimentalCallback,
+	__experimentalReactComponent,
+	__experimentalClass,
+	__experimentalVariable
+} = unlock( experiments );
+```
+
+#### Experimental function arguments
+
+To add an experimental argument to a stable function you'll need
+to prepare a stable and an experimental version of that function.
+Then, export the stable function and `lock()` the unstable function
+inside it:
+
+```js
+// In @wordpress/package1/index.js:
+import { lock } from './experiments';
+
+// The experimental function contains all the logic
+function __experimentalValidateBlocks(formula, __experimentalIsStrict) {
+	let isValid = false;
+	// ...complex logic we don't want to duplicate...
+	if ( __experimentalIsStrict ) {
+		// ...
+	}
+	// ...complex logic we don't want to duplicate...
+
+	return isValid;
+}
+
+// The stable public function is a thin wrapper that calls the
+// experimental function with the experimental features disabled
+export function validateBlocks(blocks) {
+	__experimentalValidateBlocks(blocks, false);
+}
+lock( validateBlocks, __experimentalValidateBlocks );
+
+
+// In @wordpress/package2/index.js:
+import { validateBlocks } from '@wordpress/package1';
+import { unlock } from './experiments';
+
+// The experimental function may be "unlocked" given the stable function:
+const __experimentalValidateBlocks = unlock(validateBlocks);
+__experimentalValidateBlocks(blocks, true);
+```
+
+#### Experimental React Component properties
+
+To add an experimental argument to a stable component you'll need
+to prepare a stable and an experimental version of that component.
+Then, export the stable function and `lock()` the unstable function
+inside it:
+
+```js
+// In @wordpress/package1/index.js:
+import { lock } from './experiments';
+
+// The experimental component contains all the logic
+const ExperimentalMyButton = ( { title, __experimentalShowIcon = true } ) => {
+	// ...complex logic we don't want to duplicate...
+  
+	return (
+		<button>
+			{ __experimentalShowIcon  && <Icon src={some icon} /> } { title }  
+		</button>
+	);
+}
+
+// The stable public component is a thin wrapper that calls the
+// experimental component with the experimental features disabled
+export const MyButton = ( { title } ) => 
+    <ExperimentalMyExistingButton title={ title } __experimentalShowIcon={ false } />
```

Comment:
Should `ExperimentalMyExistingButton` be `ExperimentalMyButton`? Just trying to grok the example...

Reply:
Yes it should, good spot!